File: slcm/slcm/page/exam_attendance_sheet/exam_attendance_sheet.js
frappe.pages['exam-attendance-sheet'].on_page_load = function (wrapper) {
	frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Exam Attendance Sheet',
		single_column: true
	});
	new ExamAttendanceSheet(wrapper);
};

class ExamAttendanceSheet {
	constructor(wrapper) {
		this.$body = $(wrapper).find('.page-content').css({ padding: '16px 20px' });
		this.exam_plan = null;
		this.courses = [];
		this.$toolbar = null;
		this._build_ui();
	}

	// ── UI skeleton ───────────────────────────────────────────────────────────
	_build_ui() {
		/* Filter card */
		this.$filter = $(`
			<div style="background:#fff;border:1px solid #e2e6ea;border-radius:8px;
				padding:20px 24px;margin-bottom:18px;display:flex;
				align-items:flex-end;gap:16px;flex-wrap:wrap;">
				<div style="display:flex;flex-direction:column;gap:4px;min-width:320px;">
					<label style="font-size:12px;font-weight:600;color:#495057;margin:0;">
						Exam Plan <span style="color:#e03e3e">*</span>
					</label>
					<div class="eas-plan-wrap"></div>
				</div>
				<button class="btn btn-primary btn-sm eas-load-btn"
					style="height:32px;min-width:130px;" disabled>
					<i class="fa fa-search"></i>&nbsp; Load Courses
				</button>
			</div>
		`).appendTo(this.$body);

		/* Results container — never destroyed, only emptied */
		this.$results = $('<div class="eas-results"></div>').appendTo(this.$body);

		this._build_plan_field();
	}

	// ── Exam Plan link field ──────────────────────────────────────────────────
	_build_plan_field() {
		this._plan_ctrl = frappe.ui.form.make_control({
			df: {
				fieldtype: 'Link',
				fieldname: 'exam_plan',
				options: 'Exam Plan',
				placeholder: 'Select Exam Plan…',
			},
			parent: this.$filter.find('.eas-plan-wrap')[0],
			render_input: true,
		});
		this._plan_ctrl.refresh();

		const $btn = this.$filter.find('.eas-load-btn');

		/* Enable/disable button whenever the field value changes.
		   Use both native input events AND set_value override so we catch
		   keyboard entry, mouse-select from dropdown, and programmatic sets. */
		const _sync_btn = () => {
			const v = this._plan_value();
			$btn.prop('disabled', !v);
			if (!v) {
				this.exam_plan = null;
				this._clear();
			}
		};

		this._plan_ctrl.$input.on('input change awesomplete-selectcomplete', _sync_btn);

		const _orig_set = this._plan_ctrl.set_value.bind(this._plan_ctrl);
		this._plan_ctrl.set_value = (v) => { _orig_set(v); _sync_btn(); };

		$btn.on('click', () => {
			const v = this._plan_value();
			if (!v) {
				frappe.show_alert({ message: 'Please select an Exam Plan.', indicator: 'orange' });
				return;
			}
			this.exam_plan = v;
			this._load_courses();
		});
	}

	_plan_value() {
		return (this._plan_ctrl.get_value() || this._plan_ctrl.$input.val() || '').trim();
	}

	// ── Helpers ───────────────────────────────────────────────────────────────
	_clear() {
		/* Only empties the DOM results area; does NOT touch this.courses
		   (callers that need to reset courses do so explicitly). */
		this.$results.empty();
		this.$toolbar = null;
	}

	_fmt_time(t) {
		if (!t) return '<span class="text-muted">—</span>';
		try { return moment(t, 'HH:mm:ss').format('hh:mm A'); }
		catch (e) { return t; }
	}

	// ── Load courses ──────────────────────────────────────────────────────────
	_load_courses() {
		this.courses = [];          // reset data
		this._clear();              // reset DOM

		this.$results.html(`
			<div class="text-center text-muted" style="padding:40px;">
				<i class="fa fa-spinner fa-spin fa-2x"></i>
				<p style="margin-top:12px;">Loading courses…</p>
			</div>`);

		frappe.call({
			method: 'slcm.slcm.page.exam_attendance_sheet.exam_attendance_sheet.get_exam_courses',
			args: { exam_plan: this.exam_plan },
			callback: (r) => {
				if (r.exc) { this._clear(); return; }

				const data = r.message || [];
				this._clear();   // remove spinner before rendering

				if (!data.length) {
					this.$results.html(`
						<div class="alert alert-warning" style="margin:0;">
							<b>No courses scheduled</b> in <i>"${this.exam_plan}"</i>.
							Open the Exam Plan and add course schedules first.
						</div>`);
					return;
				}

				/* Store THEN render — order matters */
				this.courses = data;
				this._render_courses();
			}
		});
	}

	// ── Render course table ───────────────────────────────────────────────────
	_render_courses() {
		/* NOTE: do NOT call _clear() here — this.courses must remain intact */

		/* Toolbar */
		this.$toolbar = $(`
			<div style="display:flex;align-items:center;gap:10px;flex-wrap:wrap;
				margin-bottom:14px;background:#fff;border:1px solid #e2e6ea;
				border-radius:8px;padding:14px 18px;">
				<button class="btn btn-success btn-sm eas-gen-all">
					<i class="fa fa-qrcode"></i>&nbsp; Generate All Barcodes
				</button>
				<button class="btn btn-primary btn-sm eas-export" disabled>
					<i class="fa fa-download"></i>&nbsp; Download Attendance Excel
				</button>
				<span class="eas-msg text-muted" style="font-size:12px;margin-left:4px;"></span>
			</div>
		`).appendTo(this.$results);

		this.$toolbar.find('.eas-gen-all').on('click', () =>
			this._generate_barcodes(this.courses.map(c => c.course)));

		this.$toolbar.find('.eas-export').on('click', () => this._export_excel());

		/* Group by date */
		const date_groups = {};
		this.courses.forEach(c => {
			const key = c.exam_date || '__nodate__';
			if (!date_groups[key]) date_groups[key] = [];
			date_groups[key].push(c);
		});

		const sorted_dates = Object.keys(date_groups).sort((a, b) => {
			if (a === '__nodate__') return 1;
			if (b === '__nodate__') return -1;
			return a < b ? -1 : 1;
		});

		sorted_dates.forEach(date => {
			const label = (date === '__nodate__')
				? '(No Date Scheduled)'
				: frappe.datetime.str_to_user(date);

			/* Date header */
			$(`<div style="background:#f0f4ff;border-left:4px solid #2490ef;
					padding:9px 16px;margin-bottom:6px;border-radius:4px;
					font-weight:600;font-size:13px;color:#1a5276;
					display:flex;align-items:center;gap:8px;">
				<i class="fa fa-calendar-o"></i> Exam Date: ${label}
			</div>`).appendTo(this.$results);

			/* Course table */
			const $card = $(`
				<div style="background:#fff;border:1px solid #e2e6ea;
					border-radius:8px;margin-bottom:16px;overflow:hidden;">
					<table class="table table-bordered" style="margin:0;font-size:13px;">
						<thead style="background:#f8f9fa;">
							<tr>
								<th style="width:36px;text-align:center;">#</th>
								<th>Course</th>
								<th style="width:100px;">Code</th>
								<th style="width:105px;">Start</th>
								<th style="width:105px;">End</th>
								<th style="width:150px;">Venue / Block</th>
								<th style="width:120px;">Hall / Room</th>
								<th style="width:110px;text-align:center;">Barcodes</th>
								<th style="width:110px;text-align:center;">Action</th>
							</tr>
						</thead>
						<tbody></tbody>
					</table>
				</div>
			`).appendTo(this.$results);

			date_groups[date].forEach((c, i) =>
				this._append_course_row(c, i + 1, $card.find('tbody')));
		});

		this._sync_export_btn();
	}

	// ── Single course row + detail row ────────────────────────────────────────
	_append_course_row(c, idx, $tbody) {
		const has = c.barcode_count > 0;

		const $tr = $(`
			<tr data-course="${c.course}">
				<td style="text-align:center;color:#6c757d;">${idx}</td>
				<td>
					<a class="eas-course-link" href="#"
						style="font-weight:600;color:#2490ef;text-decoration:none;"
						title="Click to view enrolled students">
						${c.course_name || c.course}
					</a>
					<i class="fa fa-chevron-right eas-chevron"
						style="font-size:10px;color:#adb5bd;margin-left:6px;
						transition:transform .2s;"></i>
				</td>
				<td><span class="text-muted" style="font-size:12px;">${c.course_code || c.course}</span></td>
				<td>${this._fmt_time(c.start_time)}</td>
				<td>${this._fmt_time(c.end_time)}</td>
				<td>${c.venue || '<span class="text-muted">—</span>'}</td>
				<td>${c.hall  || '<span class="text-muted">—</span>'}</td>
				<td style="text-align:center;">
					<span class="badge eas-bc-badge badge-${has ? 'success' : 'secondary'}"
						style="font-size:12px;padding:4px 8px;">
						${c.barcode_count}
					</span>
				</td>
				<td style="text-align:center;">
					<button class="btn btn-xs btn-default eas-gen-one"
						style="font-size:11px;">
						<i class="fa fa-refresh"></i> Generate
					</button>
				</td>
			</tr>
		`).appendTo($tbody);

		/* Detail/student row (hidden) */
		const $detail = $(`
			<tr class="eas-detail" data-course="${c.course}"
				style="display:none;background:#f9fbff;">
				<td colspan="9" style="padding:0;">
					<div class="eas-student-panel" style="padding:12px 20px;"></div>
				</td>
			</tr>
		`).appendTo($tbody);

		/* Toggle student list on course-name click */
		$tr.find('.eas-course-link').on('click', (e) => {
			e.preventDefault();
			const open = $detail.is(':visible');
			$detail.toggle(!open);
			$tr.find('.eas-chevron').css('transform', open ? '' : 'rotate(90deg)');
			if (!open) this._load_students(c, $detail.find('.eas-student-panel'));
		});

		/* Per-course generate */
		$tr.find('.eas-gen-one').on('click', () => this._generate_barcodes([c.course]));
	}

	// ── Student list panel ────────────────────────────────────────────────────
	_load_students(c, $panel) {
		/* If already loaded, keep existing content */
		if ($panel.children().length) return;

		$panel.html('<i class="fa fa-spinner fa-spin"></i> Loading students…');

		frappe.call({
			method: 'slcm.slcm.page.exam_attendance_sheet.exam_attendance_sheet.get_course_students',
			args: { exam_plan: this.exam_plan, course: c.course },
			callback: (r) => {
				if (r.exc) {
					$panel.html('<span class="text-danger">Failed to load students.</span>');
					return;
				}
				this._render_student_panel($panel, r.message || [], c);
			}
		});
	}

	_render_student_panel($panel, students, c) {
		if (!students.length) {
			$panel.html(`
				<div class="text-muted" style="padding:10px 0;font-size:13px;">
					<i class="fa fa-info-circle"></i>
					No students found in Student Groups for <b>${c.course_name || c.course}</b>.
					Check that students are added to a Student Group with this course.
				</div>`);
			return;
		}

		const total = students.length;
		const done  = students.filter(s => s.has_barcode).length;

		const rows = students.map((s, i) => `
			<tr>
				<td style="text-align:center;color:#6c757d;width:40px;">${i + 1}</td>
				<td style="font-weight:500;">${s.student_name}</td>
				<td style="width:130px;">${s.registration_id || '<span class="text-muted">—</span>'}</td>
				<td style="width:80px;text-align:center;">${s.section || '<span class="text-muted">—</span>'}</td>
				<td style="width:130px;text-align:center;">
					${s.barcode
						? `<span class="badge badge-success"
								style="font-size:12px;letter-spacing:1px;">${s.barcode}</span>`
						: `<span class="badge badge-secondary"
								style="font-size:11px;">Not generated</span>`}
				</td>
			</tr>`).join('');

		$panel.html(`
			<div style="display:flex;align-items:center;justify-content:space-between;
				margin-bottom:10px;flex-wrap:wrap;gap:8px;">
				<span style="font-weight:600;font-size:13px;color:#333;">
					<i class="fa fa-users"></i>&nbsp;
					${c.course_name || c.course} —
					<span class="text-muted" style="font-weight:400;">
						${total} student(s) &nbsp;|&nbsp;
						${done} barcoded &nbsp;|&nbsp;
						${total - done} pending
					</span>
				</span>
				<button class="btn btn-xs btn-success eas-gen-panel">
					<i class="fa fa-qrcode"></i> Generate Barcodes
				</button>
			</div>
			<div style="max-height:340px;overflow-y:auto;border:1px solid #e2e6ea;border-radius:6px;">
				<table class="table table-bordered table-hover"
					style="margin:0;font-size:12px;">
					<thead style="background:#f0f4ff;position:sticky;top:0;">
						<tr>
							<th style="text-align:center;">#</th>
							<th>Name</th>
							<th>ID No.</th>
							<th style="text-align:center;">Sec</th>
							<th style="text-align:center;">Barcode</th>
						</tr>
					</thead>
					<tbody>${rows}</tbody>
				</table>
			</div>`);

		$panel.find('.eas-gen-panel').on('click', () =>
			this._generate_barcodes([c.course]));
	}

	// ── Generate barcodes ─────────────────────────────────────────────────────
	_generate_barcodes(courses) {
		frappe.call({
			method: 'slcm.slcm.page.exam_attendance_sheet.exam_attendance_sheet.generate_barcodes',
			args: {
				exam_plan: this.exam_plan,
				courses: JSON.stringify(courses),
			},
			freeze: true,
			freeze_message: 'Generating barcodes…',
			callback: (r) => {
				if (r.exc) {
					frappe.show_alert({ message: 'Error generating barcodes.', indicator: 'red' });
					return;
				}
				const res = r.message || {};
				frappe.show_alert({ message: res.message || 'Done.', indicator: 'green' });
				/* Reload to refresh badge counts and clear cached student panels */
				this._load_courses();
			}
		});
	}

	// ── Export Excel ──────────────────────────────────────────────────────────
	_export_excel() {
		frappe.show_alert({ message: 'Preparing Excel…', indicator: 'blue' });
		frappe.call({
			method: 'slcm.slcm.page.exam_attendance_sheet.exam_attendance_sheet.export_attendance_excel',
			args: { exam_plan: this.exam_plan },
			freeze: true,
			freeze_message: 'Building Excel file…',
			callback: (r) => {
				if (r.exc) {
					frappe.show_alert({ message: 'Error generating Excel.', indicator: 'red' });
					return;
				}
				const { file_content, filename } = r.message;
				const bytes = Uint8Array.from(atob(file_content), c => c.charCodeAt(0));
				const blob  = new Blob([bytes], {
					type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'
				});
				const url = URL.createObjectURL(blob);
				const a = document.createElement('a');
				a.href = url; a.download = filename;
				document.body.appendChild(a); a.click();
				setTimeout(() => { URL.revokeObjectURL(url); a.remove(); }, 1000);
				frappe.show_alert({ message: 'Excel downloaded!', indicator: 'green' });
			}
		});
	}

	// ── Sync export button state ──────────────────────────────────────────────
	_sync_export_btn() {
		if (!this.$toolbar) return;
		const total = this.courses.reduce((s, c) => s + (c.barcode_count || 0), 0);
		this.$toolbar.find('.eas-export').prop('disabled', total === 0);
	}
}

File: slcm/slcm/page/exam_attendance_sheet/exam_attendance_sheet.py
# Copyright (c) 2026, TFSS and contributors
# For license information, please see license.txt

import frappe


@frappe.whitelist()
def get_exam_plans():
	"""Return list of active exam plans."""
	return frappe.db.get_all(
		"Exam Plan",
		filters={"status": "Active"},
		fields=["name", "exam_name", "term"],
		order_by="exam_name asc",
	)


@frappe.whitelist()
def get_exam_courses(exam_plan):
	"""Return courses scheduled under the given exam plan, with dates and barcode counts."""
	schedules = frappe.db.get_all(
		"Exam Course Schedule",
		filters={"parent": exam_plan, "parenttype": "Exam Plan"},
		fields=["course", "exam_date", "start_time", "end_time", "venue", "hall"],
		order_by="exam_date asc, course asc",
	)

	if not schedules:
		return []

	# Fetch course names and codes
	course_info = {
		r.name: {"course_name": r.course_name, "course_code": r.course_code or r.name}
		for r in frappe.db.get_all(
			"Course",
			filters={"name": ["in", [s["course"] for s in schedules]]},
			fields=["name", "course_name", "course_code"],
		)
	}

	# Count existing barcodes per course using raw SQL (group_by with COUNT)
	barcode_counts = {}
	rows = frappe.db.sql(
		"""
		SELECT course, COUNT(name) AS cnt
		FROM `tabExam Barcode`
		WHERE exam_plan = %s
		GROUP BY course
		""",
		(exam_plan,),
		as_dict=True,
	)
	for r in rows:
		barcode_counts[r.course] = r.cnt

	for s in schedules:
		info = course_info.get(s["course"], {})
		s["course_name"] = info.get("course_name", s["course"])
		s["course_code"] = info.get("course_code", s["course"])
		s["barcode_count"] = barcode_counts.get(s["course"], 0)

	return schedules


@frappe.whitelist()
def generate_barcodes(exam_plan, courses=None):
	"""
	Generate 6-digit unique barcodes for all students enrolled in the given
	courses under the exam plan. Existing barcodes are preserved; only missing
	students get new ones.

	courses: JSON list of course names (optional, defaults to all scheduled).
	"""
	import json
	import random
	import datetime

	if isinstance(courses, str):
		courses = json.loads(courses)

	if not courses:
		schedules = frappe.db.get_all(
			"Exam Course Schedule",
			filters={"parent": exam_plan, "parenttype": "Exam Plan"},
			fields=["course", "exam_date"],
		)
		courses = list({s.course for s in schedules})

	# Build exam_date lookup per course
	date_lookup = {}
	for s in frappe.db.get_all(
		"Exam Course Schedule",
		filters={"parent": exam_plan, "parenttype": "Exam Plan", "course": ["in", courses]},
		fields=["course", "exam_date"],
	):
		date_lookup[s.course] = s.exam_date

	# Collect all barcodes already issued under this plan (for uniqueness)
	used_barcodes = set(
		r.barcode
		for r in frappe.db.get_all("Exam Barcode", filters={"exam_plan": exam_plan}, fields=["barcode"])
	)

	def _new_barcode():
		for _ in range(10000):
			code = str(random.randint(100000, 999999))
			if code not in used_barcodes:
				used_barcodes.add(code)
				return code
		frappe.throw("Could not generate a unique 6-digit barcode. Barcode pool may be exhausted.")

	now = frappe.utils.now_datetime()
	created = 0

	for course in courses:
		students = _get_students_for_course(course, exam_plan)
		if not students:
			continue

		# Find which students already have barcodes
		existing = {
			r.student
			for r in frappe.db.get_all(
				"Exam Barcode",
				filters={"exam_plan": exam_plan, "course": course},
				fields=["student"],
			)
		}

		for st in students:
			if st["student"] in existing:
				continue
			doc = frappe.get_doc({
				"doctype": "Exam Barcode",
				"exam_plan": exam_plan,
				"course": course,
				"exam_date": date_lookup.get(course),
				"student": st["student"],
				"student_name": st["student_name"],
				"registration_id": st["registration_id"],
				"section": st["section"],
				"barcode": _new_barcode(),
				"generated_on": now,
			})
			doc.insert(ignore_permissions=True)
			created += 1

	frappe.db.commit()
	return {"created": created, "message": f"{created} new barcode(s) generated."}


@frappe.whitelist()
def get_barcodes(exam_plan, course=None):
	"""Return all barcodes for the exam plan (optionally filtered by course)."""
	filters = {"exam_plan": exam_plan}
	if course:
		filters["course"] = course

	return frappe.db.get_all(
		"Exam Barcode",
		filters=filters,
		fields=["course", "exam_date", "student", "student_name", "registration_id", "section", "barcode"],
		order_by="course asc, registration_id asc",
	)


@frappe.whitelist()
def get_course_students(exam_plan, course):
	"""
	Return students for a course with their barcode (if generated).
	Source: Student Course Marks (same as Examination Result page).
	"""
	students = _get_students_for_course(course, exam_plan)

	# Existing barcodes
	existing_barcodes = {
		r.student: r.barcode
		for r in frappe.db.get_all(
			"Exam Barcode",
			filters={"exam_plan": exam_plan, "course": course},
			fields=["student", "barcode"],
		)
	}

	result = []
	for st in students:
		result.append({
			"student": st["student"],
			"student_name": st["student_name"],
			"registration_id": st["registration_id"],
			"section": st["section"],
			"barcode": existing_barcodes.get(st["student"], ""),
			"has_barcode": st["student"] in existing_barcodes,
		})

	result.sort(key=lambda x: (x["registration_id"] or "zzz", x["student_name"]))
	return result


@frappe.whitelist()
def export_attendance_excel(exam_plan):
	"""
	Generate the exam attendance Excel with one sheet per exam date.
	Each sheet lists all courses for that date with student rows.
	Returns a base64-encoded .xlsx file.
	"""
	import io
	import base64
	import datetime
	import openpyxl
	from openpyxl.styles import Font, Alignment, Border, Side, PatternFill
	from openpyxl.utils import get_column_letter

	plan_doc = frappe.get_doc("Exam Plan", exam_plan)

	# Group schedules by exam_date
	schedules = frappe.db.get_all(
		"Exam Course Schedule",
		filters={"parent": exam_plan, "parenttype": "Exam Plan"},
		fields=["course", "exam_date", "start_time", "end_time", "venue", "hall"],
		order_by="exam_date asc, course asc",
	)

	if not schedules:
		frappe.throw("No courses scheduled in this Exam Plan.")

	# Fetch course names and codes
	course_name_map = {}
	for r in frappe.db.get_all(
		"Course",
		filters={"name": ["in", list({s["course"] for s in schedules})]},
		fields=["name", "course_name", "course_code"],
	):
		label = r.course_name or r.name
		if r.course_code:
			label = f"{label} [{r.course_code}]"
		course_name_map[r.name] = label

	# Fetch all barcodes for this plan
	barcodes = frappe.db.get_all(
		"Exam Barcode",
		filters={"exam_plan": exam_plan},
		fields=["course", "exam_date", "student", "student_name", "registration_id", "section", "barcode"],
		order_by="registration_id asc",
	)
	# Map: course -> sorted list of student rows
	barcode_map = {}
	for b in barcodes:
		barcode_map.setdefault(b.course, []).append(b)

	# Get institution info from Academic Term if available
	institution = "Institution"
	term_label = ""
	try:
		if plan_doc.term:
			term_doc = frappe.get_doc("Academic Term", plan_doc.term)
			term_label = term_doc.term_name or plan_doc.term
			if hasattr(term_doc, "academic_year"):
				yd = frappe.db.get_value("Academic Year", term_doc.academic_year, "year_name") or term_doc.academic_year
				term_label = f"{term_label}, AY {yd}"
	except Exception:
		term_label = plan_doc.term or ""

	# Group schedules by date
	from collections import defaultdict
	date_groups = defaultdict(list)
	for s in schedules:
		date_groups[s.exam_date].append(s)

	wb = openpyxl.Workbook()
	wb.remove(wb.active)  # remove default sheet

	# Styles
	thin = Side(style="thin")
	border = Border(left=thin, right=thin, top=thin, bottom=thin)

	def _hdr_font(size=12, bold=False):
		return Font(name="Bookman Old Style", size=size, bold=bold)

	def _cell_font():
		return Font(name="Bookman Old Style", size=11)

	def _center(cell):
		cell.alignment = Alignment(horizontal="center", vertical="center", wrap_text=True)

	def _set_col_widths(ws, widths):
		for i, w in enumerate(widths, 1):
			ws.column_dimensions[get_column_letter(i)].width = w

	sorted_dates = sorted(date_groups.keys(), key=lambda d: (d or ""))

	for exam_date in sorted_dates:
		date_str = ""
		if exam_date:
			try:
				date_str = frappe.utils.formatdate(str(exam_date), "dd/MM/yyyy")
			except Exception:
				date_str = str(exam_date)

		sheet_name = date_str.replace("/", "-") if date_str else "No Date"
		ws = wb.create_sheet(title=sheet_name[:31])

		day_schedules = date_groups[exam_date]
		current_row = 1

		for sched in day_schedules:
			course = sched.course
			course_display = course_name_map.get(course, course)
			students = sorted(barcode_map.get(course, []), key=lambda x: (x.registration_id or ""))

			# ── Header block ────────────────────────────────────────────────
			# Row 1: Institution name
			ws.merge_cells(
				start_row=current_row, start_column=1,
				end_row=current_row, end_column=8
			)
			c = ws.cell(row=current_row, column=1, value=institution)
			c.font = _hdr_font(14, bold=True)
			_center(c)
			ws.row_dimensions[current_row].height = 22
			current_row += 1

			# Row 2: Term / Year info
			ws.merge_cells(
				start_row=current_row, start_column=1,
				end_row=current_row, end_column=8
			)
			c = ws.cell(row=current_row, column=1, value=term_label)
			c.font = _hdr_font(12, bold=True)
			_center(c)
			ws.row_dimensions[current_row].height = 18
			current_row += 1

			# Row 3: Course + Date
			exam_line = f"{course_display} End Term Examination  ({date_str})"
			ws.merge_cells(
				start_row=current_row, start_column=1,
				end_row=current_row, end_column=8
			)
			c = ws.cell(row=current_row, column=1, value=exam_line)
			c.font = _hdr_font(11, bold=True)
			_center(c)
			ws.row_dimensions[current_row].height = 16
			current_row += 1

			# Row 4: Column headers
			col_headers = [
				"Sl.No", "Name", "ID no.", "Sec", "Barcode",
				"Answer Booklet\nnumbers", "Student\nSignature",
				"No. of Additional\nBooklet Used"
			]
			fill = PatternFill(start_color="D9E1F2", end_color="D9E1F2", fill_type="solid")
			for col_idx, header in enumerate(col_headers, 1):
				c = ws.cell(row=current_row, column=col_idx, value=header)
				c.font = Font(name="Bookman Old Style", size=11, bold=True)
				c.alignment = Alignment(horizontal="center", vertical="center", wrap_text=True)
				c.border = border
				c.fill = fill
			ws.row_dimensions[current_row].height = 35
			current_row += 1

			# Rows 5+: Student data
			if students:
				for sl, st in enumerate(students, 1):
					row_data = [
						sl,
						st.student_name or "",
						st.registration_id or "",
						st.section or "",
						st.barcode or "",
						"",  # Answer Booklet numbers (blank for faculty)
						"",  # Student Signature
						"",  # No. of Additional Booklet Used
					]
					for col_idx, val in enumerate(row_data, 1):
						c = ws.cell(row=current_row, column=col_idx, value=val)
						c.font = _cell_font()
						c.border = border
						if col_idx in (1, 3, 4, 5, 6, 8):
							_center(c)
						else:
							c.alignment = Alignment(horizontal="left", vertical="center")
					ws.row_dimensions[current_row].height = 25
					current_row += 1
			else:
				# No barcodes generated yet - placeholder row
				ws.merge_cells(
					start_row=current_row, start_column=1,
					end_row=current_row, end_column=8
				)
				c = ws.cell(row=current_row, column=1,
							value="No barcodes generated for this course yet.")
				c.font = Font(name="Bookman Old Style", size=11, italic=True, color="FF0000")
				_center(c)
				current_row += 1

			# Gap between courses on same date
			current_row += 2

		# Set column widths (once per sheet, applies to all courses)
		_set_col_widths(ws, [6, 28, 14, 8, 10, 16, 16, 16])

	# Save to buffer
	buf = io.BytesIO()
	wb.save(buf)
	buf.seek(0)
	file_b64 = base64.b64encode(buf.read()).decode()
	filename = f"Exam_Attendance_{exam_plan.replace(' ', '_')}.xlsx"
	return {"file_content": file_b64, "filename": filename}


# ── Internal helpers ────────────────────────────────────────────────────────────

def _get_students_for_course(course, exam_plan):
	"""
	Return list of dicts {student, student_name, registration_id, section}
	sourced from Student Course Marks — the same table used by Examination Result.
	Students are sorted by registration_id ascending.
	"""
	rows = frappe.db.sql(
		"""
		SELECT
			scm.student,
			CONCAT_WS(' ', sm.first_name, sm.last_name) AS student_name,
			COALESCE(NULLIF(sm.registration_id, ''), sm.name) AS registration_id,
			COALESCE(cc.section, '') AS section
		FROM `tabStudent Course Marks` scm
		LEFT JOIN `tabStudent Master` sm ON sm.name = scm.student
		LEFT JOIN (
			SELECT cs.student, cc2.section, cc2.course
			FROM `tabClass Student` cs
			INNER JOIN `tabClass Configuration` cc2 ON cc2.name = cs.parent
		) cc ON cc.student = scm.student AND cc.course = scm.course
		WHERE scm.exam_plan = %(exam_plan)s
		  AND scm.course = %(course)s
		ORDER BY COALESCE(NULLIF(sm.registration_id, ''), sm.name) ASC
		""",
		{"exam_plan": exam_plan, "course": course},
		as_dict=True,
	)

	# Deduplicate — a student could appear in multiple class configurations
	seen = set()
	result = []
	for r in rows:
		if r.student in seen:
			continue
		seen.add(r.student)
		result.append({
			"student": r.student,
			"student_name": r.student_name or r.student,
			"registration_id": r.registration_id or "",
			"section": r.section or "",
		})

	return result
